### ISSUE TITLE
CROSS JOIN confuses parser?

### ISSUE THREAD

**Issue by osvenskan** [sqlfluff/sqlfluff#871]
## Expected Behaviour
SQLFluff should recognize the alias `foo`

## Observed Behaviour

The alias `foo` isn't recognized. When I remove the `CROSS JOIN`, SQLFluff no longer emits L026 for `foo`. 

```
$ sqlfluff lint test.sql
== [test.sql] FAIL
L:   7 | P:   5 | L026 | Reference 'foo.name' refers to table/view 'foo' not
                       | found in the FROM clause or found in parent subquery.
L:   8 | P:   5 | L026 | Reference 'foo.value' refers to table/view 'foo' not
                       | found in the FROM clause or found in parent subquery.
L:  12 | P:   1 | L003 | Indent expected and not found compared to line #10
L:  12 | P:  16 | L031 | Avoid using aliases in join condition
L:  12 | P:  20 | L032 | Found USING statement. Expected only ON statements.
```

## Steps to Reproduce
`sqlfluff lint test.sql` with default config

Contents of `test.sql` --
```sql
WITH constants AS (
    SELECT 8760 AS hours_per_year
)

SELECT
    table1.name,
    foo.name,
    foo.value * constants.hours_per_year AS some_value
FROM table1
CROSS JOIN
    constants
JOIN table2 AS foo USING (id)
;
```

## Version
```
$ sqlfluff --version
sqlfluff, version 0.4.1
$ python --version
Python 3.9.2
```

## Configuration
I don't have a config file; I'm using the default config for 0.4.1.

**Comment by osvenskan** [comment]
When I downgrade to 0.4.0, linting the same file doesn't emit L026 for `foo`, so perhaps this is a regression in 0.4.1?

```
$ sqlfluff version
0.4.0
 $ sqlfluff lint test.sql
== [test.sql] FAIL
L:  12 | P:  16 | L031 | Avoid using aliases in join condition
L:  12 | P:  20 | L032 | Found USING statement. Expected only ON statements.
```

**Comment by barrywhart** [comment]
Below is a partial parse tree for the test query in 0.4.1.
```
[162](1, 10, 1)     |                        join_clause:
[162](1, 10, 1)     |                            keyword:                          'CROSS'
[167](1, 10, 6)     |                            whitespace:                       ' '
[168](1, 10, 7)     |                            keyword:                          'JOIN'
[172](1, 10, 11)    |                            [META] indent:
[172](1, 10, 11)    |                            newline:                          '\n'
[173](1, 11, 1)     |                            whitespace:                       '    '
[177](1, 11, 5)     |                            table_expression:
[177](1, 11, 5)     |                                [META] indent:
[177](1, 11, 5)     |                                main_table_expression:
[177](1, 11, 5)     |                                    table_reference:
[177](1, 11, 5)     |                                        identifier:           'constants'
[186](1, 11, 14)    |                                newline:                      '\n'
[187](1, 12, 1)     |                                [META] dedent:
[187](1, 12, 1)     |                                join_clause:
[187](1, 12, 1)     |                                    keyword:                  'JOIN'
[191](1, 12, 5)     |                                    [META] indent:
[191](1, 12, 5)     |                                    whitespace:               ' '
[192](1, 12, 6)     |                                    table_expression:
[192](1, 12, 6)     |                                        [META] indent:
[192](1, 12, 6)     |                                        main_table_expression:
[192](1, 12, 6)     |                                            table_reference:
[192](1, 12, 6)     |                                                identifier:   'table2'
[198](1, 12, 12)    |                                        whitespace:           ' '
[199](1, 12, 13)    |                                        alias_expression:
[199](1, 12, 13)    |                                            keyword:          'AS'
[201](1, 12, 15)    |                                            whitespace:       ' '
[202](1, 12, 16)    |                                            identifier:       'foo'
[205](1, 12, 19)    |                                        [META] dedent:
[205](1, 12, 19)    |                                    whitespace:               ' '
[206](1, 12, 20)    |                                    keyword:                  'USING'
[211](1, 12, 25)    |                                    [META] indent:
[211](1, 12, 25)    |                                    whitespace:               ' '
[212](1, 12, 26)    |                                    start_bracket:            '('
[213](1, 12, 27)    |                                    [META] indent:
[213](1, 12, 27)    |                                    identifier:               'id'
[215](1, 12, 29)    |                                    [META] dedent:
[215](1, 12, 29)    |                                    end_bracket:              ')'
[216](1, 12, 30)    |                                    [META] dedent:
[216](1, 12, 30)    |                                    [META] dedent:
[216](1, 12, 30)    |                            [META] dedent:
```

Here is the same area of the parse tree for the same query in version 0.4.0:
```
[162](1, 10, 1)     |                    join_clause:
[162](1, 10, 1)     |                        keyword:                              'CROSS'
[167](1, 10, 6)     |                        whitespace:                           ' '
[168](1, 10, 7)     |                        keyword:                              'JOIN'
[172](1, 10, 11)    |                        [META] indent:
[172](1, 10, 11)    |                        newline:                              '\n'
[173](1, 11, 1)     |                        whitespace:                           '    '
[177](1, 11, 5)     |                        table_expression:
[177](1, 11, 5)     |                            main_table_expression:
[177](1, 11, 5)     |                                table_reference:
[177](1, 11, 5)     |                                    identifier:               'constants'
[186](1, 11, 14)    |                        [META] dedent:
[186](1, 11, 14)    |                    newline:                                  '\n'
[187](1, 12, 1)     |                    join_clause:
[187](1, 12, 1)     |                        keyword:                              'JOIN'
[191](1, 12, 5)     |                        [META] indent:
[191](1, 12, 5)     |                        whitespace:                           ' '
[192](1, 12, 6)     |                        table_expression:
[192](1, 12, 6)     |                            main_table_expression:
[192](1, 12, 6)     |                                table_reference:
[192](1, 12, 6)     |                                    identifier:               'table2'
[198](1, 12, 12)    |                            whitespace:                       ' '
[199](1, 12, 13)    |                            alias_expression:
[199](1, 12, 13)    |                                keyword:                      'AS'
[201](1, 12, 15)    |                                whitespace:                   ' '
[202](1, 12, 16)    |                                identifier:                   'foo'
[205](1, 12, 19)    |                        whitespace:                           ' '
[206](1, 12, 20)    |                        keyword:                              'USING'
[211](1, 12, 25)    |                        [META] indent:
[211](1, 12, 25)    |                        whitespace:                           ' '
[212](1, 12, 26)    |                        start_bracket:                        '('
[213](1, 12, 27)    |                        [META] indent:
[213](1, 12, 27)    |                        identifier:                           'id'
[215](1, 12, 29)    |                        [META] dedent:
[215](1, 12, 29)    |                        end_bracket:                          ')'
[216](1, 12, 30)    |                        [META] dedent:
[216](1, 12, 30)    |                        [META] dedent:
```

Note that in 0.4.1, `JOIN table2` is nested **underneath** the `CROSS JOIN`, while in 0.4.0, they were at the same level of the parse tree. So I agree, this is a regression, and I think the nesting explains why the `foo` alias is not "visible" to the `SELECT`.

**Comment by alanmcruickshank** [comment]
this was much harder than expected but I think I've resolved it. PR coming shortly. @barrywhart - I'd love your review on it.
